draft-ietf-ppm-dap.md

```diff
@@ -1377,21 +1377,26 @@ This message consists of:
 * `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
   previous step.
 
-Finally, the Leader sends a PUT request to
-`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The payload
-is set to `AggregationJobInitReq` and the media type is set to
+Finally, the Leader sends a request to
+`PUT {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The
+payload is set to `AggregationJobInitReq` and the media type is set to
 "application/dap-aggregation-job-init-req".
 
 The Leader MUST authenticate its requests to the Helper using a scheme that
 meets the requirements in {{request-authentication}}.
 
+After receiving HTTP status 201 Created, the Leader proceeds to poll the
```

Comment:
What the recommended way of polling here? Do we poll each job in batch sequentially until all jobs give expected response? This would be blocking operation on leader side and generates noisy request to the helper. It would be better to provide ability to query for completed jobs. This enables leader to process-and-fetch the completed jobs in a loop.

Reply:
Aggregation jobs are independent, so it shouldn't be necessary to process jobs sequentially. Rather the leader can/should eagerly send aggregate init/continue to the helper.

> This would be blocking operation on leader side

Indeed, but it shouldn't be much worse than blocking leader-side waiting for the aggregate init/continue request to finish.

I'd actually expect this grants leader implementations more flexibility in how they schedule aggregation job processing. If a leader-side worker thread is blocked on an open HTTP request, it can't set it aside to do other work. In the proposed model, the leader-side worker can poll once, see that the aggregation job isn't ready, move on to other work (e.g. polling other workers), and poll again later.

> generates noisy request to the helper

Indeed that's a tradeoff, but I'm hoping that empty GET requests are fairly cheap in terms of bandwidth. And aggregator pairings are free to tune polling intervals and aggregation job volume such that the amount of noise is low.

> It would be better to provide ability to query for completed jobs. This enables leader to process-and-fetch the completed jobs in a loop.

That's an interesting idea. Can you elaborate on the structure of such an endpoint?

Reply:
I think one challenge for "give me any complete job" is deciding when to have the Helper _stop_ returning a given job to this request type.

If the Helper returns a given job only once, an inopportune network issue, Leader crash, or other transient issue could lead to the job being "lost". (specifically, the condition for a job being "lost" would be the Helper considering the request successful, but the Leader not processing the response successfully)

If the Helper hands out the job until e.g. the Leader sends the next `AggregationJobContinueReq`, a Leader which sends multiple requests for completed jobs concurrently could see the same response multiple times. (Also, if we have reached the last aggregation step, there is no next `AggregationJobContinueReq` to be sent.)

Reply:
Something to note is that async init/continue raises interesting possibilities as to how many reports are contained within an `AggregationJobInitReq`.

Right now, the leader/helper have to limit the number of reports based on how long HTTP timeouts are configured helper-side.

With async, the leader can send arbitrarily large aggregation jobs without concerns about the HTTP gateway timeout. If taking advantage of that, there are fewer aggregation jobs that require bookkeeping and polling.

Reply:
Resolving this subthread: I feel strongly that per-job polling is sufficient, and that implementations can control the volume of GET polls by tweaking the aggregation job size and polling interval.

Feel free to unresolve if there are dissents.